### DISCUSSION TITLE
How to initialize an integrator instance by a scene's parameters?

### DISCUSSION THREAD

**Discussion by Mephisto405** [mitsuba-renderer/mitsuba3#507] [Q&A] [answered]
Let's say I'm writing a new integrator that requires to do some heavy computation relying on the scene parameters (e.g., bbox, shapes, bsdfs) once before rendering. This computation is only required once per scene, and every later rendering does not require this computation but just needs to query the precomputed result. Thus, I'd like to compute the value in the __init__() function. That means, as far as I think, I need to get the scene parameters from the input of __init__() function, that is `mi.Properties()`. But how...?
Any help would be very appreciated.

The below code is an example.
```python
class MyIntegrator(mi.ad.common.ADIntegrator):

    def __init__(
        self, props=mi.Properties()
    ):
        super().__init__(props)

        self.scene = # HOW???
        m_area = []
        for shape in scene.shapes():
            m_area.append(shape.surface_area())
        m_area = np.array(m_area)[:, 0]
        self.m_area = m_area

    def sample(...):
        var1 = self.m_area[...]
```

**Comment by njroussel** [comment, marked as answer]
Hi @Mephisto405 

Plugins in Mitsuba have a unidirectional parent->child relationship, which makes something like you're trying to do a bit complicated as usually the scene is the parent and the integrator is one of its children. There's not really a way to (easily) get around this.

There are a few options to work around it though. The issue here is that the `__init__` constructor cannot take the `mi.Scene` it is a part of as an argument, however you can use any other method or even create a new method which takes a scene as a parameter.
A few options:

1. Add a new `initialize(scene: mi.Scene)` method to your integrator which you need to explicitly call.
2. Make `sample()` check if  it already has initialized the pre-computed data or if there is a new scene, if not then it can compute it.

Basically, I don't think there is any _built-in mitsuba way_ of achieving this cleanly. You would need to abuse some of the existing interface.

**Reply by Mephisto405** [reply]
I see. Actually, yesterday I tried the first method you suggested, but I thought I ruined Mitsuba's AdIntegrator interface, and it made me feel uncomfortable. But I realized there is no built-in Mitsuba way of solution, so I think it will be more comfortable to implement it according to that method. Thanks for your quick reply, as always.

**Comment by Mephisto405** [comment]
Very similar case: https://github.com/mitsuba-renderer/mitsuba3/blob/master/src/python/python/ad/integrators/prbvolpath.py
